src/lib.rs

```diff
@@ -1178,4 +1259,70 @@ mod tests {
         assert_eq!(hue, hue_after_mixing(Color::graytone(0.7)));
         assert_eq!(hue, hue_after_mixing(Color::white()));
     }
+
+    #[test]
+    fn color_scale_add_preserves_ordering() {
+        let mut color_scale = ColorScale::empty();
+
+        color_scale
+            .add_stop(Color::red(), Fraction::from(0.5))
+            .add_stop(Color::gray(), Fraction::from(0.0))
+            .add_stop(Color::blue(), Fraction::from(1.0));
+
+        assert_eq!(color_scale.color_stops.get(0).unwrap().color, Color::gray());
+        assert_eq!(color_scale.color_stops.get(1).unwrap().color, Color::red());
+        assert_eq!(color_scale.color_stops.get(2).unwrap().color, Color::blue());
+    }
+
+    #[test]
+    fn color_scale_empty_sample_none() {
+        let mix: Box<dyn Fn(&Color, &Color, Fraction) -> Color> =
+            Box::new(|c1: &Color, c2: &Color, f: Fraction| c1.mix::<Lab>(c2, f));
+
+        let color_scale = ColorScale::empty();
+
+        let color = color_scale.sample(Fraction::from(0.0), &mix);
+
+        assert_eq!(color, None);
+    }
+
+    #[test]
+    fn color_scale_sample() {
+        let mix: Box<dyn Fn(&Color, &Color, Fraction) -> Color> =
+            Box::new(|c1: &Color, c2: &Color, f: Fraction| c1.mix::<Lab>(c2, f));
+
+        let mut color_scale = ColorScale::empty();
+
+        color_scale
+            .add_stop(Color::red(), Fraction::from(0.0))
+            .add_stop(Color::green(), Fraction::from(1.0));
+
+        let sample_red_green = color_scale.sample(Fraction::from(0.5), &mix).unwrap();
+
+        let mix_red_green = mix(&Color::red(), &Color::green(), Fraction::from(0.5));
+
+        assert_eq!(sample_red_green, mix_red_green);
+    }
+
+    #[test]
+    fn color_scale_sample_position() {
+        let mix: Box<dyn Fn(&Color, &Color, Fraction) -> Color> =
+            Box::new(|c1: &Color, c2: &Color, f: Fraction| c1.mix::<Lab>(c2, f));
+
+        let mut color_scale = ColorScale::empty();
+
+        color_scale
+            .add_stop(Color::red(), Fraction::from(0.0))
+            .add_stop(Color::green(), Fraction::from(0.5))
+            .add_stop(Color::blue(), Fraction::from(1.0));
+
+        let sample_red_green = color_scale.sample(Fraction::from(0.25), &mix).unwrap();
+        let sample_green_blue = color_scale.sample(Fraction::from(0.75), &mix).unwrap();
+
+        let mix_red_green = mix(&Color::red(), &Color::green(), Fraction::from(0.50));
+        let mix_green_blue = mix(&Color::green(), &Color::blue(), Fraction::from(0.50));
```

Comment:
can we please also add tests for sampling at `0.0`, `0.5` and `1.0`?

what would happen if two colors are added at 0.5?

Reply:
I think it's better to replace the color stop when adding a color at the same position. I'll also add a test for this.